workflow for patching k8s workloads

ydayagi opened the pull request:
FLPATH-2204
https://issues.redhat.com/browse/FLPATH-2204

ydayagi commented:
@masayag @pkliczewski @gabriel-farache

masayag commented:
Pls add README.md for the workflow, see https://github.com/rhdhorchestrator/serverless-workflows/blob/main/workflows/create-ocp-project/README.md as example.

Is it expected to run the workflow from the orchestrator plugin? in that case, shouldn't you include also the workflow-output.json as part of it so the response can be set in the UI as the result message of the workflow?

ydayagi commented on pipeline/workflow-builder.Dockerfile:
```diff
@@ -13,7 +13,7 @@ FROM ${BUILDER_IMAGE:-registry.redhat.io/openshift-serverless-1/logic-swf-builde
 # ARG QUARKUS_EXTENSIONS=org.kie:kogito-addons-quarkus-jobs-knative-eventing:999-SNAPSHOT,org.kie:kie-addons-quarkus-persistence-jdbc:999-SNAPSHOT,io.quarkus:quarkus-jdbc-postgresql:3.8.4,io.quarkus:quarkus-agroal:3.8.4,org.kie:kie-addons-quarkus-monitoring-prometheus:999-SNAPSHOT,org.kie:kie-addons-quarkus-monitoring-sonataflow:999-SNAPSHOT
 # When using prod:
-ARG QUARKUS_EXTENSIONS=org.kie:kie-addons-quarkus-monitoring-sonataflow:10.0.0,org.kie:kogito-addons-quarkus-jobs-knative-eventing:10.0.0,org.kie:kie-addons-quarkus-persistence-jdbc:10.0.0,io.quarkus:quarkus-jdbc-postgresql:3.8.6.redhat-00004,io.quarkus:quarkus-agroal:3.8.6.redhat-00004
+ARG QUARKUS_EXTENSIONS=io.quarkiverse.openapi.generator:quarkus-openapi-generator:2.9.1-lts,org.kie:kie-addons-quarkus-monitoring-sonataflow:10.0.0,org.kie:kogito-addons-quarkus-jobs-knative-eventing:10.0.0,org.kie:kie-addons-quarkus-persistence-jdbc:10.0.0,io.quarkus:quarkus-jdbc-postgresql:3.8.6,io.quarkus:quarkus-agroal:3.8.6
```
did u build the images for the workflows lately? won't work. this is after troubleshooting with ricardo. issue fixed in this version.

gabriel-farache replied:
> did u build the images for the workflows lately?

No, hence my question but we did build it in the demo repo and we do not do that cc @masayag something to keep in mind maybe

ydayagi replied:
perhaps u used old images and did not rebuild images in the demo repo. perhaps you did not need to pass a token to a server. if you do need to set a token in application.properties then u need this extension. I would gladly get reid of it.

ydayagi commented on workflows/patch-k8s-resource/patch-k8s-resource.sw.yaml:
```diff
@@ -0,0 +1,77 @@
+        transition: "SetApiVersion3"
+    defaultCondition:
+      transition: "SetApiVersion1"
+  - name: SetApiVersion1
```
the name has no meaning . it is simply used to separate into 3 states. You think anyone will understand what SetApisAppsV1 means? I doubt it.

ydayagi replied:
not everything. but in this case it makes sense cause there's on meaningful name for the state. a name that a new person would understand. I'll repeat my question: "You think anyone will understand what SetApisAppsV1 means?"

ydayagi replied:
this is a workflow to be consumed by other apps only

ydayagi commented on workflows/patch-k8s-resource/patch-k8s-resource.sw.yaml:
```diff
@@ -0,0 +1,77 @@
+          arguments:
+            apiVersion: .resourceApiVersion
+            namespace: .resourceNamespace
+            kind: '.resourceType+"s"'
```
this worklfow is not just a general use workflow. it is for the optimizations plugin. in general you could say that it is for applying redhat insights recommendations. there the kind/type is singular.
However, UX wise it is also better this way for the user. we never use the plural name (deployments, pods etc). we use the singular name. However, the kube clients require the plural name in the API.

ydayagi commented on workflows/patch-k8s-resource/schemas/patch-k8s-resource.json:
```diff
@@ -0,0 +1,78 @@
+          "properties": {
+            "cpu": {
+              "title": "cpu",
+              "type": "number"
```
cpu is indeed float/double. user can not input 500m. it is not a number. intuitive or not is debatable. to me 0.5 is more intuitive then 500m. do u say 'give me half a glass of coke' or do you say 'give me 500m of a glass of coke'?. the fact is that this is the default format in k8s and redhta insights. we need it and prefer it that way

gabriel-farache replied:
In K8s/OCP in general the unit is used in the CPU instead of using decimal.

> we need it and prefer it that way

who is "we"?

ydayagi replied:
optimization plugin and my optimizer app. this is a requirement we had. this workflow is for these apps.

ydayagi replied:
we are not delivering this workflow as a general purpose workflow.

ydayagi commented on workflows/patch-k8s-resource/schemas/patch-k8s-resource.json:
```diff
@@ -0,0 +1,78 @@
+            },
+            "memory": {
+              "title": "memory",
+              "type": "integer"
```
we do not want to support all memory formats. only bytes. this is our decision

gabriel-farache replied:
from UX using bytes is wrong, and as you want to ease the user with not using plural form of the resource type, I think the effort should be consistent and applied here as well

who is "we"?

ydayagi replied:
optimization plugin and my optimizer app. this is a requirement we had. this workflow is for these apps.

ydayagi replied:
we are not delivering this workflow as a general purpose workflow.

ydayagi commented:
> Pls add README.md for the workflow, see https://github.com/rhdhorchestrator/serverless-workflows/blob/main/workflows/create-ocp-project/README.md as example.
> 
> Is it expected to run the workflow from the orchestrator plugin? in that case, shouldn't you include also the workflow-output.json as part of it so the response can be set in the UI as the result message of the workflow?

readme? i honestly don't understand what is the use of it. the other readmes gave me no information about the workflows.
what is the use of this output json? I worked without it and everything was fine. please provide example

gabriel-farache commented:
> what is the use of this output json? I worked without it and everything was fine. please provide example

The output is used by the RHDH UI plugin to display things nicely, it is applied in all workflows

And why readmes? Because they explain what the workflow is doing, what are the expected env vars and inputs

ydayagi commented:
> Pls add README.md for the workflow, see https://github.com/rhdhorchestrator/serverless-workflows/blob/main/workflows/create-ocp-project/README.md as example.
> 
> Is it expected to run the workflow from the orchestrator plugin? in that case, shouldn't you include also the workflow-output.json as part of it so the response can be set in the UI as the result message of the workflow?

I think I should have given some background. this a workflow to be consumed by Optimizations plugin. I will add it to the workflow description and README. @gabriel-farache FYI

ydayagi commented on workflows/patch-k8s-resource/patch-k8s-resource.sw.yaml:
```diff
@@ -0,0 +1,77 @@
+#events:
+#  - name: fromPooling
+#    type: org.acme.sw.camel.pooling
+#    source: urn:camel-pooling
```
fixed

ydayagi commented on workflows/patch-k8s-resource/patch-k8s-resource.sw.yaml:
```diff
@@ -0,0 +1,77 @@
+name: patch-k8s-resource
+description: patch-k8s-resource
+dataInputSchema: schemas/patch-k8s-resource.json
+#start: WaitForEvent
```
fixed

ydayagi commented:
@masayag @gabriel-farache 
I added readme and removed commented code. as I explain in the README, this workflow is only to be used by Optimizations plugin and Optimizer app. Not for humans.

ydayagi commented on workflows/patch-k8s-resource/patch-k8s-resource.sw.yaml:
```diff
@@ -0,0 +1,77 @@
+  #    onEvents:
+  #      - eventRefs:
+  #          - fromPooling
+  #    transition: SetApiVersion
```
fixed
